### PR TITLE
Jl/dictionary

### PR DESCRIPTION
**Pull request by jbieniosek**
Adding a first pass of the dictionary lesson

### CODE REVIEW

File: 00-preparing-for-ada/dictionaries.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
+# Dictionaries
+
+## Learning Goals
+
+At the end of this lesson students will be able to:
```
Comment (CheezItMan) [inline 00-preparing-for-ada/dictionaries.md:5]: Just something I've seen from Simon's units, leaving off the sentence at the beginning and just list the LGs. 

```suggestion
```

File: 00-preparing-for-ada/dictionaries.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+### Why Dictionaries?
+
+In the lists lesson we used our first complex data structure.  Lists are powerful and useful tools, but they do have some limitations.  Say we want to use our list to store all of the pieces of an address.  We could of course use several variables to store the city, state, zip code, street address, but as we saw in the list lesson, being able to store all of the information in one variable can make it simpler to access the data and pass the data to functions.  If we use a list, we can store each of these pieces of information, but to access them we need to remember the index where we stored them.  We could set up constants like STATE_INDEX to help make this code easier to read, but it's not ideal.  Enter dictionaries!  Dictionaries associate each piece of data with a unique key, and then we can use the key to access the data in the dictionary.  In the case of our address data structure, we can just make the keys meaningful words like "city" and "state", and then in our code whenever we want to access these pieces of the address, we can use the appropriate key to grab the data we want from the dictionary.
```
Comment (CheezItMan) [inline 00-preparing-for-ada/dictionaries.md:19]:
```suggestion
In the lists lesson we used our first complex data structure.  Lists are powerful and useful tools, but they do have some limitations.  Say we want to use our list to store all of the pieces of an address.  We could of course use several variables to store the city, state, zip code, street address, but as we saw in the list lesson, being able to store all of the information in one variable can make it simpler to access the data and pass the data to functions.  If we use a list, we can store each of these pieces of information, but to access them we need to remember the index where we stored them.  

We could set up constants like:

```python
STATE_INDEX = 0
CITY_INDEX = 1

print( f"We live in {address[CITY_INDEX]}, {address[STATE_INDEX]}.")
```

Using constants to remember the indices would code easier to read, but it's still not ideal.  Enter dictionaries!  

Dictionaries associate each piece of data with a unique key, and then we can use the key to access the data in the dictionary.  In the case of our address data structure, we can just make the keys meaningful words like "city" and "state", and then in our code whenever we want to access these pieces of the address, we can use the appropriate key to grab the data we want from the dictionary.  
```

File: 00-preparing-for-ada/dictionaries.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+### Why Dictionaries?
+
+In the lists lesson we used our first complex data structure.  Lists are powerful and useful tools, but they do have some limitations.  Say we want to use our list to store all of the pieces of an address.  We could of course use several variables to store the city, state, zip code, street address, but as we saw in the list lesson, being able to store all of the information in one variable can make it simpler to access the data and pass the data to functions.  If we use a list, we can store each of these pieces of information, but to access them we need to remember the index where we stored them.  We could set up constants like STATE_INDEX to help make this code easier to read, but it's not ideal.  Enter dictionaries!  Dictionaries associate each piece of data with a unique key, and then we can use the key to access the data in the dictionary.  In the case of our address data structure, we can just make the keys meaningful words like "city" and "state", and then in our code whenever we want to access these pieces of the address, we can use the appropriate key to grab the data we want from the dictionary.  
+
+## Vocabulary and Synonyms
```
Comment (CheezItMan) [inline 00-preparing-for-ada/dictionaries.md:21]: I do suggest using a table for vocab, like what [simon has done](https://github.com/Ada-Developers-Academy/core-unit-1/blob/main/intro-to-git/collaborating.md#vocabulary-and-synonyms)

File: 00-preparing-for-ada/dictionaries.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
...We have the word and we have the list of correctly guessed letters.  Our end goal is to display each letter of the word with an '_' character if the letter has not yet been guessed and the correct letter if it has been guessed.  Every time a new correct letter is guessed, we want to update this display with the new information that we have.  We could use loops and lists to solve this problem by looping over each letter in the word and checking to see if it's in the list of guessed letters and based on a conditional test displaying either the letter or the underscore character.  There is an algorithmic cost to solving the problem this way.  Every time we search through our list it takes time, and we're repeating the same search each time we want to display our word.  A way to limit the number of times we do these searches is to save the result of the search.  We are going to use dictionaries to do that!  We are going to build a dictionary for each letter that contains the letter and a "guessed" value that we can set to indicate that the user has guessed the letter.  We will store all of the dictionaries in a list and use the resulting list to construct our word display for our user.
```
Comment (CheezItMan) [inline 00-preparing-for-ada/dictionaries.md:46]: Just a bit wordy.

```suggestion
The last piece of our snowman game is displaying to the user the letters they have guessed correctly.  We have:

 -  The word 
 -  The list of correctly guessed letters  
 
Our end goal is to display each letter of the word with an '_' character if the letter has not yet been guessed and the correct letter if it has been guessed.  

Every time a new correct letter is guessed, we want to update this display with the new information that we have.  We could use loops and lists to solve this problem by looping over each letter in the word and checking to see if it's in the list of guessed letters and based on a conditional test displaying either the letter or the underscore character.  There is an algorithmic cost to solving the problem this way.  

Every time we search through our list it takes time, and we're repeating the same search each time we want to display our word.  A way to limit the number of times we do these searches is to save the result of the search.  We are going to use dictionaries to do that!  

We are going to build a dictionary for each letter that contains the letter and a "guessed" value that we can set to indicate that the user has guessed the letter.  We will store all of the dictionaries in a list and use the resulting list to construct our word display for our user.
```

I would also either use simpler language or define "algorithmic cost."

Maybe just use a bulleted list of the steps if you were going to search for each guessed letter each time it needed to display the guessed letters in the word.

File: 00-preparing-for-ada/dictionaries.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+### Updating the List of Dictionaries
+
+Right now our `print_word_list` will always just print out underscores because we're not yet updating the list of dictionaries when our user guesses a correct letter.  The next step is to write a function that takes a letter and the list of dictionaries and updates the guessed boolean for each dictionary that contains that letter.  We are going to make this function do a little extra work.  We are also going to ask this function to return `True` or `False` to tell us if all of the letters have been guessed or not.  We already have to loop through the list to compare our letter to the letter value of each dictionary, so we can use the same loop to also check to see if all of the guessed values are `True`.  We will call the function `update_and_check_word_list`.
+
+This function will:
+* Accept a list of dictionaries and a letter as parameters
+* Set the 'guessed' value to `True` for each dictionary who's 'letter' value matches the letter parameter
+* Return `True` if all of the letters have been guessed and `False` if they have not
+
+__[TODO - PYTHON CODE CHALLENGE HERE!]__
```
Comment (CheezItMan) [inline 00-preparing-for-ada/dictionaries.md:176]: I'll create a PR with the python challenge shortly.

File: 00-preparing-for-ada/dictionaries.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+### Ending the Game
+
+Our game currently ends when our user maxes out their wrong guesses, but we also want to end the game if our user guesses all of the letters in the word.  We also want to display messages to our user if they lose or win.
+
+Our `snowman` function needs to be updated in the following ways:
+* Add a function call to `update_and_check_word_list` when the user guesses a correct letter
+* Store the return variable from the above function call and use it to end the game loop if the user has guessed all of the letters
+* After the game loop, print "Congratulations, you win!" if our user won or "Sorry, you lose!  The word was ..." with the secret word replacing ... if the user lost.
+
+__[TODO - PYTHON CODE CHALLENGE HERE!]__
```
Comment (CheezItMan) [inline 00-preparing-for-ada/dictionaries.md:187]: @jbieniosek 
Can you add an HTML comment here with the starter code you want students to have?

File: 00-preparing-for-ada/functions.md [hunk 7]
```diff
@@ -14,7 +14,7 @@ In this section we will be building on the code that you wrote in the previous l
 
 ### Why Functions?
```
Comment (CheezItMan) [inline 00-preparing-for-ada/functions.md:15]: Just a note, it's cleaner if you make the corrections in individual branches.   So keep this branch focused on dictionaries.

File: 00-preparing-for-ada/dictionaries.md (hunk 4 above)
Comment (kaidamasaki) [inline 00-preparing-for-ada/dictionaries.md:46]:
```suggestion
The final piece of our Snowman game is showing the user the letters guessed in the word so far.  Our goal is to show the word to the user with an `_` instead of each letter they have yet to guess.  To do this we will keep track of a list of correctly guessed letters. Then every time a new correct letter is guessed we will update our display.  

We could loop over each letter in the word to compare against the list every time but this is more work than we need to do.  We can be clever and store the results of doing this check in a data structure so we only have to check against our list once per new guess.  We are going to dictionaries to do that!  

We are going to build a list with little dictionary for each letter that contains a "letter" key and a "guessed" key that indicates whether or not the user has guessed it already.  We will store all of the dictionaries in a list and use the resulting list to construct our word display for our user.
```

**Review by CheezItMan — COMMENTED** [review]
I reviewed the dictionaries part and gave some feedback.  I'll add 1 python challenge shortly, and if you can give me input on what you'd like the starter code for students to look like in the 2nd python challenge I'll get that in as well.

**Review by kaidamasaki — COMMENTED** [review]
Well, looks like Chris reviewed this while I was writing but here are my notes for the beginning.

File: 00-preparing-for-ada/dictionaries.md [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
+# Adding a new key-value pair to an existing dictionary
+new_dict["key3"] = "value3"
+
+# Accessing a key-value pair
+value = new_dict["key1"]
+
+```
+
+* key: A string.  All of the keys in a dictionary must be unique.
```
[the remark this answers is not in the document]
Comment (jbieniosek) [reply 00-preparing-for-ada/dictionaries.md:41]: I checked w/ Audrey before I wrote this section and the new curriculum is going teach keys as strings and is not going to go into immutable vs mutable data

File: 00-preparing-for-ada/dictionaries.md (hunk 6 above)
Comment (jbieniosek) [reply 00-preparing-for-ada/dictionaries.md:187]: I'm not totally sure, I would like them to have:
def snowman():

and then they put in their snowman code, but there are several helper functions for snowman that they've written over the previous lessons.  The end goal is that their snowman code looks something like:

def snowman():
    r = RandomWord()
    snowman_word = r.word(word_min_length=SNOWMAN_MIN_WORD_LENGTH, word_max_length=SNOWMAN_MAX_WORD_LENGTH)
    #print(f"debug info: {snowman_word}")
    snowman_list = build_word_list(snowman_word)
    correct_guesses_list = []
    wrong_guesses_list = []
    all_guessed = False
    print_word_list(snowman_list)
    while len(wrong_guesses_list) < SNOWMAN_WRONG_GUESSES and not all_guessed:
        user_input = get_letter_from_user(correct_guesses_list, wrong_guesses_list)
        if user_input in snowman_word:
            print("You guessed a letter that's in the word!")
            correct_guesses_list.append(user_input)
            all_guessed = update_and_check_word_list(snowman_list, user_input)
        else:
            print(f"The letter {user_input} is not in the word")
            wrong_guesses_list.append(user_input)
        print_snowman_graphic(len(wrong_guesses_list))
        print_word_list(snowman_list)
        print("Wrong guesses: " + " ".join(wrong_guesses_list))

    if all_guessed:
        print("Congratulations, you win!")
    else:
        print(f"Sorry, you lose!  The word was {snowman_word}")

So they have written build_word_list, print_word_list, update_and_check_word_list, get_letter_from_user and print_snowman_graphic and they should use all of those in this function.  The the loops lesson walked them through adding the RandomWord package, so I'm guessing that needs to be mocked.

**Review by kaidamasaki — APPROVED** [review]

File: example_code/games_dictionary.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
+import random
+from wonderwords import RandomWord
+#https://pypi.org/project/wonderwords/
```
Comment (CheezItMan) [inline example_code/games_dictionary.py:3]: Have you taught students how to add a package?

File: 00-preparing-for-ada/dictionaries.md (hunk 6 above)
Comment (CheezItMan) [reply 00-preparing-for-ada/dictionaries.md:396]: Where did RandomWord come from, did you teach students how to add a package?

File: example_code/games_dictionary.py (hunk 9 above)
Comment (jbieniosek) [reply example_code/games_dictionary.py:3]: This section is added in the lists lesson
